lib/blockstore/file.js

```diff
@@ -0,0 +1,579 @@
+const {types, prefixes} = require('./common');
+
+/**
+ * File Block Store
+ *
+ * @alias module:blockstore:FileBlockStore
+ * @abstract
+ */
+
+class FileBlockStore extends AbstractBlockStore {
+  /**
+   * Create a blockstore that stores blocks in files.
+   * @constructor
+   */
+
+  constructor(options) {
+    super(options);
+
+    assert(isAbsolute(options.location), 'Location not absolute.');
+
+    this.location = options.location;
+    this.indexLocation = resolve(this.location, './index');
+
+    this.db = bdb.create({
+      location: this.indexLocation,
+      cacheSize: options.cacheSize,
+      compression: false
+    });
+
+    this.maxFileLength = options.maxFileLength || 128 * 1024 * 1024;
+
+    assert(Number.isSafeInteger(this.maxFileLength),
+      'Invalid max file length.');
+
+    this.network = Network.primary;
+
+    if (options.network != null)
+      this.network = Network.get(options.network);
+
+    this.writing = false;
+  }
+
+  /**
+   * Compares the number of files in the directory
+   * with the recorded number of files.
+   * @param {Number} type - The type of block data
+   * @private
+   * @returns {Promise}
+   */
+
+  async check(type) {
+    const prefix = prefixes[type];
+    const regexp = new RegExp(`^${prefix}(\\d{5})\\.dat$`);
+    const all = await fs.readdir(this.location);
+    const dats = all.filter(f => regexp.test(f));
+    const filenos = dats.map(f => parseInt(f.match(regexp)[1]));
+
+    let missing = false;
+
+    for (const fileno of filenos) {
+      const rec = await this.db.get(layout.f.encode(type, fileno));
+      if (!rec) {
+        missing = true;
+        break;
+      }
+    }
+
+    return {missing, filenos};
+  }
+
+  /**
+   * Creates indexes from files for a block type. Reads the hash of
+   * the block data from the magic prefix, except for a block which
+   * the hash is read from the block header.
+   * @private
+   * @param {Number} type - The type of block data
+   * @returns {Promise}
+   */
+
+  async _index(type) {
+    const {missing, filenos} = await this.check(type);
+
+    if (!missing)
+      return;
+
+    this.logger.info('Indexing block type %d...', type);
+
+    for (const fileno of filenos) {
+      const b = this.db.batch();
+      const filepath = this.filepath(type, fileno);
+      const data = await fs.readFile(filepath);
+      const reader = bio.read(data);
+      let magic = null;
+      let blocks = 0;
+
+      while (reader.left() >= 4) {
+        magic = reader.readU32();
+
+        // Move forward a byte from the last read
+        // if the magic doesn't match.
+        if (magic !== this.network.magic) {
+          reader.seek(-3);
+          continue;
+        }
+
+        let hash = null;
+        let position = 0;
+        let length = 0;
+
+        try {
+          length = reader.readU32();
+
+          if (type === types.BLOCK) {
+            position = reader.offset;
+            const header = Headers.fromReader(reader);
```

Comment:
Seems like we could optimize this with `hash256.digest(reader.readBytes(80, true))`.

Reply:
Yep, works. Though, it may not have a huge impact as headers are small and not that many of them.

Reply:
Added